lib/logstash/outputs/elasticsearch.rb

```diff
@@ -273,68 +270,86 @@ def register
       common_options.merge(:hosts => @hosts, :logger => @logger)
     )
 
-    if @manage_template
-      begin
-        @logger.info("Automatic template management enabled", :manage_template => @manage_template.to_s)
-        @client.template_install(@template_name, get_template, @template_overwrite)
-      rescue => e
-        @logger.error("Failed to install template: #{e.message}")
-      end
-    end
+    TemplateManager.install_template(self)
 
     @logger.info("New Elasticsearch output", :hosts => @hosts)
 
-    @client_idx = 0
+    @buffer = ::LogStash::Outputs::ElasticSearch::Buffer.new(@logger, @flush_size, @idle_flush_time) do |actions|
+      retrying_submit(actions)
+    end
+  end # def register
 
-    buffer_initialize(
-      :max_items => @flush_size,
-      :max_interval => @idle_flush_time,
-      :logger => @logger
-    )
+  def receive(event)
+    params = event_action_params(event)
+    action = event.sprintf(@action)
 
-    @retry_timer_thread = Thread.new do
-      loop do
-        sleep(@retry_max_interval)
-        @retry_flush_mutex.synchronize { @retry_queue_needs_flushing.signal }
-      end
-    end
+    @buffer << [action, params, event]
+  end # def receive
 
-    @retry_thread = Thread.new do
-      while @retry_close_requested.false?
-        @retry_flush_mutex.synchronize { @retry_queue_needs_flushing.wait(@retry_flush_mutex) }
-        retry_flush
+  # When there are exceptions raised upon submission, we raise an exception so that
+  # Stud::Buffer will retry to flush
+  def flush
+    @buffer.flush
+  end # def flush
+
+  def close
+    @stopping.make_true
+    @client.stop_sniffing!
+    @buffer.stop
+  end
+
+  def retrying_submit(actions)
+    retries_left = @max_retries+1 # +1 for the first attempt
+
+    # Initially we submit the full list of actions
+    submit_actions = actions
+
+    while submit_actions && submit_actions.length > 0 && retries_left > 0
+      return if !submit_actions # If everything's a success we move along
+      # We retry with whatever is didn't succeed
+      begin
+        submit_actions = submit(submit_actions)
+      rescue => e
+        @logger.warn("Encountered an unexpected error submitting a bulk request!",
```

Comment:
Do we still log the individual requests that failed and its corresponding ES errors?